ticker/src/main/java/com/robinhood/ticker/LevenshteinUtils.java

```diff
@@ -12,26 +30,124 @@
+        while (true) {
+            // Check for terminating conditions
+            final boolean reachedEndOfSource = sourceIndex == source.length;
+            final boolean reachedEndOfTarget = targetIndex == target.length;
+            if (reachedEndOfSource && reachedEndOfTarget) {
+                break;
+            } else if (reachedEndOfSource) {
+                final int[] remainingActions = new int[target.length - targetIndex];
+                Arrays.fill(remainingActions, ACTION_INSERT);
+                columnActions.add(remainingActions);
+                totalNumActions += remainingActions.length;
+                break;
+            } else if (reachedEndOfTarget) {
+                final int[] remainingActions = new int[source.length - sourceIndex];
+                Arrays.fill(remainingActions, ACTION_DELETE);
+                columnActions.add(remainingActions);
+                totalNumActions += remainingActions.length;
+                break;
+            }
+
+            final boolean containsSourceChar = supportedCharacters.contains(source[sourceIndex]);
+            final boolean containsTargetChar = supportedCharacters.contains(target[targetIndex]);
+
+            if (containsSourceChar && containsTargetChar) {
+                // We reached a segment that we can perform animations on
+                final int sourceEndIndex =
+                        findNextUnsupportedChar(source, sourceIndex + 1, supportedCharacters);
+                final int targetEndIndex =
+                        findNextUnsupportedChar(target, targetIndex + 1, supportedCharacters);
+                final int[] columnActionsForSegment = computeColumnActionsForSegment(
+                        source, target, sourceIndex, sourceEndIndex, targetIndex, targetEndIndex
+                );
+                columnActions.add(columnActionsForSegment);
+                totalNumActions += columnActionsForSegment.length;
+                sourceIndex = sourceEndIndex;
+                targetIndex = targetEndIndex;
+            } else if (containsSourceChar) {
+                // We are animating in a target character that isn't supported
+                columnActions.add(new int[] { ACTION_INSERT });
+                totalNumActions++;
+                targetIndex++;
+            } else if (containsTargetChar) {
+                // We are animating out a source character that isn't supported
+                columnActions.add(new int[] { ACTION_DELETE });
+                totalNumActions++;
+                sourceIndex++;
+            } else {
+                // Both characters are not supported, perform default animation to replace
+                columnActions.add(new int[] { ACTION_SAME });
+                totalNumActions++;
+                sourceIndex++;
+                targetIndex++;
+            }
+        }
+
+        // Concat all of the actions into one array
+        final int[] result = new int[totalNumActions];
```

Comment:
this whole bottom section seems a little inefficient. is there no way to allocate this whole array at the top and insert directly into it? what if it were a List<Integer> instead to make it easier to insert/shift?